src/hotspot/share/runtime/threadHeapSampler.hpp

```diff
@@ -27,23 +27,59 @@
 #define SHARE_RUNTIME_THREADHEAPSAMPLER_HPP
 
 #include "memory/allocation.hpp"
+#include "oops/oopsHierarchy.hpp"
 
 class ThreadHeapSampler {
  private:
-  size_t _bytes_until_sample;
+  // Amount of bytes to allocate before taking the next sample
+  size_t _sample_threshold;
+
+  // The TLAB top address when the last sampling happened, or
+  // TLAB start if a new TLAB is allocated
+  HeapWord* _tlab_top_at_sample_start;
+
+  // The accumulated amount of allocated bytes in a TLAB since the last sampling
+  // excluding the amount between _tlab_sample_start and top
+  size_t _accumulated_tlab_bytes_since_sample;
+
+  // The accumulated amount of allocated bytes outside TLABs since last sample point
+  size_t _accumulated_outside_tlab_bytes_since_sample;
+
   // Cheap random number generator
   static uint64_t _rnd;
 
   static volatile int _sampling_interval;
 
   void pick_next_geometric_sample();
-  void pick_next_sample(size_t overflowed_bytes = 0);
+  void pick_next_sample();
 
   static double fast_log2(const double& d);
   uint64_t next_random(uint64_t rnd);
 
+  size_t unsampled_in_current_tlab(HeapWord* tlab_top)  const {
+    // Both can be nullptr if there's not active TLAB, but otherwise
+    // they both should be non-null.
+    assert((tlab_top != nullptr) == (_tlab_top_at_sample_start != nullptr),
+           "Both should either be uninitialized or initialized "
+           "tlab_top: " PTR_FORMAT " _tlab_top_at_sample_start: " PTR_FORMAT,
+           p2i(tlab_top), p2i(_tlab_top_at_sample_start));
+
+    return pointer_delta(tlab_top, _tlab_top_at_sample_start, 1);
+  }
+
+  size_t tlab_bytes_since_last_sample(HeapWord* tlab_top) const {
+    return _accumulated_tlab_bytes_since_sample + unsampled_in_current_tlab(tlab_top);
+  }
+
+  size_t outside_tlab_bytes_since_last_sample() const {
+    return _accumulated_outside_tlab_bytes_since_sample;
+  }
+
  public:
-  ThreadHeapSampler() {
+  ThreadHeapSampler() :
+      _tlab_top_at_sample_start(nullptr),
+      _accumulated_tlab_bytes_since_sample(0),
+      _accumulated_outside_tlab_bytes_since_sample(0) {
```

Comment:
Nit: Would it make sense to initialize `_sample_threshold` as well to make it consistent?

Reply:
To me it makes sense. I didn't do it here because it gets initialized further down and I didn't want reviewers to get hung up on that change. But I'll take this as a go-ahead to make that change.